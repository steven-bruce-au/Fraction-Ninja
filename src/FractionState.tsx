
/** Type used to by the FractionComparator to describe the state of a fraction column
*/
export class FractionState {
  numerator: number
  denominator: number
  charts: number // number of whole charts required to render this fraction 
  value: number 

  constructor(numerator: number=1, denominator: number=1) {
  	this.numerator = numerator;
  	this.denominator = denominator;
    this.value = numerator / denominator;
    this.charts = Math.ceil(numerator / denominator);
  }

}