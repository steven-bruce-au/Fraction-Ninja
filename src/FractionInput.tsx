import React from 'react';
import { FractionState } from './FractionState';
import styles from './FractionInput.module.scss'

/** Type used by the FractionInput props */
export interface FractionInputProps {
  fractionState: FractionState;
    numeratorUpdated(numerator: number): void;
    denominatorUpdated(denominator: number): void;
}


/** Render the value of the fraction numerator and denominator along with the buttons to increment and decrement them
*/
export default class FractionInput extends React.Component<FractionInputProps, {}> {

  constructor(props: FractionInputProps) {
    super(props);
    this.state = {
    };

    this.handleNumeratorChange = this.handleNumeratorChange.bind(this);
    this.handleDenominatorChange = this.handleDenominatorChange.bind(this);

  }

  /** Handler for the denominator value changing, used by both increment and decrement
  @param the new denominator value
  @return a callback for the button component to trigger the change
  */
  handleDenominatorChange(denominator: number) : (event: any) => void {
    return (event: any) => {
      this.props.denominatorUpdated(denominator);
    }
  }

 /** Handler for the numerator value changing, used by both increment and decrement
  @param the new numerator value
  @return a callback for the button component to trigger the change
  */
  handleNumeratorChange(numerator: number) : (event: any) => void {
    return (event: any) => {
       this.props.numeratorUpdated(numerator);
    }
  }

  render() : JSX.Element {
    const { numerator, denominator} = this.props.fractionState;

    return (
       <div className={styles.fraction}>
         <div className={styles.fractionRow} >
          <button className={styles.button} onClick={this.handleNumeratorChange(numerator - 1)}> - </button>
          <div className={`${styles.number} ${styles.underline} ${styles.numerator}`}> { numerator }</div>
          <button className={styles.button} onClick={this.handleNumeratorChange(numerator + 1)}> + </button>
        </div>

         <div className={styles.fractionRow} >
          <button className={styles.button} onClick={this.handleDenominatorChange(denominator - 1)}>-</button>
          <div className={`${styles.number} ${styles.denominator}`}> {denominator}</div>
          <button className={styles.button} onClick={this.handleDenominatorChange(denominator + 1)}>+</button>
        </div>
       </div>
    );
  }
}
