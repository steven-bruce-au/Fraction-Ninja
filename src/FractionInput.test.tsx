import React from 'react';
import ReactDOM from 'react-dom';
import * as renderer from 'react-test-renderer';
import { FractionState } from './FractionState';
import FractionInput from './FractionInput';
import util from 'util';


it('render FractionInput', () => {

  const fractionState = new FractionState(1,3);

  // Mock functions
  const handleNumeratorChange = jest.fn();
  const handleDenominatorChange = jest.fn();

  const component = renderer.create(<FractionInput  fractionState={fractionState} 
                   numeratorUpdated={handleNumeratorChange} 
                   denominatorUpdated={handleDenominatorChange} /> );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // console.log(util.inspect(tree, {showHidden: false, depth: null}));

  // expect neither our numerator or denominator to be updated
  expect(handleNumeratorChange).not.toHaveBeenCalled();
  expect(handleDenominatorChange).not.toHaveBeenCalled();

});

it('decrement numerator', () => {

  const fractionState = new FractionState(1,3);

  const handleNumeratorChange = jest.fn();
  const handleDenominatorChange = jest.fn();

  const component = renderer.create(<FractionInput  fractionState={fractionState} 
                   numeratorUpdated={handleNumeratorChange} 
                   denominatorUpdated={handleDenominatorChange} /> );

  let tree = component.toJSON();

  // console.log(util.inspect(tree, {showHidden: false, depth: null}));

  // get fraction row
  const numerator  = tree!.children![0] as renderer.ReactTestRendererJSON;
  // get decrement button
  const decrement  = numerator!.children![0] as renderer.ReactTestRendererJSON;

  // click on decrement numerator button
  decrement.props.onClick();

  // expect our numerator to been reduced to zero but no the denominator
  expect(handleNumeratorChange).toHaveBeenCalledWith(0);
  expect(handleDenominatorChange).not.toHaveBeenCalled();

});


it('increment numerator', () => {

  const fractionState = new FractionState(1,3);

  const handleNumeratorChange = jest.fn();
  const handleDenominatorChange = jest.fn();

  const component = renderer.create(<FractionInput  fractionState={fractionState} 
                   numeratorUpdated={handleNumeratorChange} 
                   denominatorUpdated={handleDenominatorChange} /> );

  let tree = component.toJSON();

  // console.log(util.inspect(tree, {showHidden: false, depth: null}));

  // get fraction row
  const numerator  = tree!.children![0] as renderer.ReactTestRendererJSON;
  // get increment button
  const increment  = numerator!.children![2] as renderer.ReactTestRendererJSON;

  // click on increment numerator button
  increment.props.onClick();

  // expect our numerator to been reduced to zero but no the denominator
  expect(handleNumeratorChange).toHaveBeenCalledWith(2);
  expect(handleDenominatorChange).not.toHaveBeenCalled();

});


it('decrement denominator', () => {

  const fractionState = new FractionState(1,3);

  const handleNumeratorChange = jest.fn();
  const handleDenominatorChange = jest.fn();

  const component = renderer.create(<FractionInput  fractionState={fractionState} 
                   numeratorUpdated={handleNumeratorChange} 
                   denominatorUpdated={handleDenominatorChange} /> );

  let tree = component.toJSON();

  // console.log(util.inspect(tree, {showHidden: false, depth: null}));

  // get fraction row
  const denominator  = tree!.children![1] as renderer.ReactTestRendererJSON;
  // get decrement button
  const decrement  = denominator!.children![0] as renderer.ReactTestRendererJSON;

  // click on decrement numerator button
  decrement.props.onClick();

  // expect our numerator to been reduced to zero but no the denominator
  expect(handleNumeratorChange).not.toHaveBeenCalled();
  expect(handleDenominatorChange).toHaveBeenCalledWith(2);

});


it('increment denominator', () => {

  const fractionState = new FractionState(1,3);

  const handleNumeratorChange = jest.fn();
  const handleDenominatorChange = jest.fn();

  const component = renderer.create(<FractionInput  fractionState={fractionState} 
                   numeratorUpdated={handleNumeratorChange} 
                   denominatorUpdated={handleDenominatorChange} /> );

  let tree = component.toJSON();

  // console.log(util.inspect(tree, {showHidden: false, depth: null}));

  // get fraction row
  const denominator  = tree!.children![1] as renderer.ReactTestRendererJSON;
  // get increment button
  const increment  = denominator!.children![2] as renderer.ReactTestRendererJSON;

  // click on increment numerator button
  increment.props.onClick();

  // expect our numerator to been reduced to zero but no the denominator
   expect(handleNumeratorChange).not.toHaveBeenCalled();
  expect(handleDenominatorChange).toHaveBeenCalledWith(4);

});
