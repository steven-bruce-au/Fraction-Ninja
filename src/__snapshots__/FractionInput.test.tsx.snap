// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render FractionInput 1`] = `
<div
  className="fraction"
>
  <div
    className="fractionRow"
  >
    <button
      className="button"
      onClick={[Function]}
    >
       - 
    </button>
    <div
      className="number underline numerator"
    >
       
      1
    </div>
    <button
      className="button"
      onClick={[Function]}
    >
       + 
    </button>
  </div>
  <div
    className="fractionRow"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      -
    </button>
    <div
      className="number denominator"
    >
       
      3
    </div>
    <button
      className="button"
      onClick={[Function]}
    >
      +
    </button>
  </div>
</div>
`;
